internal/buildsign/ocpbuild/ocpbuildmanager.go

```diff
@@ -0,0 +1,173 @@
+)
+
+type Status string
+
+const (
+	ocpbuildTypeBuild = "build"
+	ocpbuildTypeSign  = "sign"
+
+	StatusCompleted  Status = "completed"
+	StatusCreated    Status = "created"
+	StatusInProgress Status = "in progress"
+	StatusFailed     Status = "failed"
+
+	hashAnnotation = "kmm.node.kubernetes.io/last-hash"
+)
+
+var ErrNoMatchingBuild = errors.New("no matching Build")
+
+//go:generate mockgen -source=ocpbuildmanager.go -package=ocpbuild -destination=mock_ocbuildmanager.go
+
+type ocpbuildManager interface {
+	isOCPBuildChanged(existingBuild *buildv1.Build, newBuild *buildv1.Build) (bool, error)
+	getModuleOCPBuildByKernel(ctx context.Context, modName, namespace, kernelVersion, ocbuildType string, owner metav1.Object) (*buildv1.Build, error)
+	getModuleOCPBuilds(ctx context.Context, modName, modNamespace, ocbuildType string, owner metav1.Object) ([]buildv1.Build, error)
+	getOCPBuildStatus(build *buildv1.Build) (Status, error)
+	createOCPBuild(ctx context.Context, build *buildv1.Build) error
+	deleteOCPBuild(ctx context.Context, build *buildv1.Build) error
+	ocpbuildLabels(modName, kernelVersion, ocpbuildType string) map[string]string
+}
+
+type ocpbuildManagerImpl struct {
+	client client.Client
+}
+
+func newOCPBuildManager(client client.Client) ocpbuildManager {
+	return &ocpbuildManagerImpl{
+		client: client,
+	}
+}
+
+func (omi *ocpbuildManagerImpl) isOCPBuildChanged(existingBuild *buildv1.Build, newBuild *buildv1.Build) (bool, error) {
+	existingAnnotations := existingBuild.GetAnnotations()
+	newAnnotations := newBuild.GetAnnotations()
+	if existingAnnotations == nil {
+		return false, fmt.Errorf("annotations are not present in the existing build %s", existingBuild.Name)
+	}
+	if existingAnnotations[hashAnnotation] == newAnnotations[hashAnnotation] {
+		return false, nil
+	}
+	return true, nil
+}
+
+func (omi *ocpbuildManagerImpl) getModuleOCPBuildByKernel(ctx context.Context,
+	modName,
+	modNamespace,
+	kernelVersion,
+	ocbuildType string,
+	owner metav1.Object) (*buildv1.Build, error) {
+	buildList := buildv1.BuildList{}
+
+	opts := []client.ListOption{
+		client.MatchingLabels(omi.ocpbuildLabels(modName, kernelVersion, ocbuildType)),
+		client.InNamespace(modNamespace),
+	}
+
+	if err := omi.client.List(ctx, &buildList, opts...); err != nil {
+		return nil, fmt.Errorf("could not list Build: %v", err)
+	}
+
+	// filter OCP builds by owner, since they could have been created by the preflight
+	// when checking that specific module
+	moduleOwnedOCPBuilds := filterOCPBuildsByOwner(buildList.Items, owner)
+
+	if n := len(moduleOwnedOCPBuilds); n == 0 {
+		return nil, ErrNoMatchingBuild
+	} else if n > 1 {
+		return nil, fmt.Errorf("expected 0 or 1 Builds, got %d", n)
+	}
+
+	return &moduleOwnedOCPBuilds[0], nil
+}
+
+func (omi *ocpbuildManagerImpl) getModuleOCPBuilds(ctx context.Context, modName, modNamespace, ocpbuildType string, owner metav1.Object) ([]buildv1.Build, error) {
```

Comment:
This is the same method as the one before excluding 1 kernel label. Can me use the same method with a list of labels as arg or something similar instead of duplicating the code?

Reply:
They are small functions and signature is not the same: one returns a pointer to a single Build, while the other returns a list of Builds. There is a common code (List), but moving it to a separate function does not reduce the number of lines dramatically. In upstream we have the same functions also